Headless renderer

jdrotleff opened the pull request:
Added a headless renderer class for rendering without a graphical interface. 
For the best result consider to export an appropriate view matrix for your visualization before rendering.

This class can be used analogously to the Visualizer class. An example that renders a whole simulation with given particles, vector fields in their respective lists into an output folders using a given view matrix could look like this.

```
renderer = znvis.Headless_Renderer(particle_list, vector_fields_list, out_folder, view_matrix)
renderer.render_visualization()
```

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Full camera suite (base, static, trajectory, circular, zooming, keyframe, particle-following), keyframe interpolation UI, headless visualizer, and VideoManager; cameras and headless visualizer now exposed at package top level.

* **Tests**
  * Extensive new unit and integration tests covering cameras, trajectories, visualizers, particles, and video manager.

* **Chores**
  * .gitignore updated to ignore *.mp4 and add .DS_Store.

* **Refactor**
  * Particle API accepts per-frame data; material and mesh constructors updated to accept material parameters.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

jdrotleff committed:
Added headless rendering

This class offers the option to render without a gui. Works analogously to the Visualizer class.

Currently only a static camera perspective is implemented that can be defined with the view_matrix.

jdrotleff committed:
Merge branch 'zincware:main' into headless-renderer

jdrotleff committed:
Fixed Formatting

jdrotleff committed:
Changed Class name

from Headless_Visualizer to Headless_Renderer

SamTov commented on znvis/visualizer/headless_renderer.py:
```diff
@@ -0,0 +1,284 @@
+        renderer_resolution: list = [4096, 2160],
+        renderer_spp: int = 64,
+        renderer: Mitsuba = Mitsuba(),
+        view_matrix: np.ndarray = np.array([[ 1,  0,  0, -100],
```
Can we make this optional and add some classes to deal with different approaches?

SamTov reviewed (changes requested):
I would like to have the camera options included in this PR if possible. That way it all works from day one. We also need a test.

jdrotleff committed:
Add .DS_Store files from MacOS to .gitignore

jdrotleff committed:
This update enhances control options of the camera during the rendering process.

Following Camera Options are now available

Can be used as a static camera. Needs either a view matrix or the three vectors eye, center and up.
↑ (up)
|
|       camera looks from the eye to the center
|
●------>●
eye   center
- eye: The position of the camera
- center: The position at which the camera is looking.
- up: This vector defines the orientation of the camera.

Can be used as
```
base_camera = znvis.Camera(center, eye, up)
```
Also provides methods that allow
- ```get_view_matrix_from_particle_positions(first_frame_positions)``` focusing on the center of the simulation box
- ```get_view_matrix_from_box_size(box_size)``` focusing on the center of all of particles in the simulation.
- ```get_minimal_view_matrix(box_size, rendering_resolution)``` experimental method that should find the camera setting that minimizes dead zones in your rendered image, where nothing happens.

Follows a particle during a full simulation shifted from the particle via a defined ```camera_particle_vector```.
Can be instantiated as for example
```
 vis.ParticleFollowingCamera(particle_positions=positions_of_particle_1,
		 camera_particle_vector=np.array([0,0,50]),
		 camera_up_vector=np.array([0,1,0]))
```

By passing a keyframe camera to the visualizer, you can manually stick camera positions to a certain frame.
The keyframe camera will then be able to interpolate a smooth camera trajectory between all your chosen frames.
This is exported to your output path from where you can pass it to the headless visualizer that will then render your desired trajectory.

At the moment you will need to use two scripts or this workaround

```
import subprocess

print("Starting interactive visualization...")
p1 = subprocess.Popen(["python", "vis1.py")

p1.wait()
print("Starting rendering process...")
subprocess.run(["python", "vis2.py")
```

vis1.py should include:
```
keyframe_camera = vis.KeyframeCamera()
visualizer = vis.Visualizer(particles=particle_list,
                output_folder=out_folder,
                keyframe_camera=keyframe_camera)
visualizer.run_visualization()
```

vis2.py should include :
```
keyframe_camera = vis.KeyframeCamera(view_matrices_path='./output_folder/interpolated_view_matrices.npy')

visualizer2 = vis.Headless_Visualizer(particles=particle_list,
                                    output_folder=out_folder,
                                    camera=keyframe_camera,)

visualizer2.render_visualization()
```

Allows defining a subclass ```trajectory``` object, which provides a trajectory the camera should follow through a scene.
Can be used as
```
trajectory_camera = vis.TrajectoryCamera(total_frames=number_of_frames,
						trajectory=trajectory_object)

```

Provides methods that return center, eye and up vector depending on the frame index depending on the total number of frames to create smooth trajectories the camera can follow.
At the moment the following trajectories are implemented
- Circular Trajectory
Creates a circular trajectory with ```radius```around the ```axis```-axis focusing on the ```center```. Optionally, the trajectory can be looped, ping-ponged or limited on a certain angle range. The ```frames_per_rotation``` variable defines the speed of the rotation.
An example would be:
```
circular_trajectory = vis.CircularTrajectory(total_frames=number_of_frames,
							center=np.array([Lx/2, Ly/2, 0]),
							radius=200,
							frames_per_rotation=20,
							axis='y',
							clockwise=True,
							loop=False,
							angle_range=(-np.pi/4, np.pi/4),
							ping_pong=True)
```

- Zooming Trajectory
Creates a zooming in or zooming out trajectory from ```center``` and ```initial_eye``` based on the sign of ```zoom_distance```. The ```frames_per_zoom``` argument defines the ratio during which zooming happens.
An example would be:
```
zoom_trajectory = vis.ZoomingTrajectory(total_frames=number_of_frames,
				center=np.array([Lx/2, Ly/2, 0]),
				initial_eye=np.array([Lx/2, Ly/2, 200]),
				zoom_distance = 150,
				frames_per_zoom=1.0)
```

jdrotleff committed:
Removed wrong docstring

jpd-de commented:
## Introduced new Camera Class

This update enhances control options of the camera during the rendering process.

Following Camera Options are now available

### BaseCamera
Can be used as a static camera. Needs either a view matrix or the three vectors eye, center and up.
↑ (up)
|
|       camera looks from the eye to the center
|
●------>●
eye   center
- eye: The position of the camera
- center: The position at which the camera is looking.
- up: This vector defines the orientation of the camera.

Can be used as 
```
base_camera = znvis.Camera(center, eye, up)
```
Also provides methods that allow 
- ```get_view_matrix_from_particle_positions(first_frame_positions)``` focusing on the center of the simulation box
- ```get_view_matrix_from_box_size(box_size)``` focusing on the center of all of particles in the simulation.
- ```get_minimal_view_matrix(box_size, rendering_resolution)``` experimental method that should find the camera setting that minimizes dead zones in your rendered image, where nothing happens.


### Particle Following Camera
Follows a particle during a full simulation shifted from the particle via a defined ```camera_particle_vector```. 
Can be instantiated as for example
```
 vis.ParticleFollowingCamera(particle_positions=positions_of_particle_1,
		 camera_particle_vector=np.array([0,0,50]),
		 camera_up_vector=np.array([0,1,0]))
```

### Keyframe Camera
By passing a keyframe camera to the visualizer, you can manually stick camera positions to a certain frame.
The keyframe camera will then be able to interpolate a smooth camera trajectory between all your chosen frames.
This is exported to your output path from where you can pass it to the headless visualizer that will then render your desired trajectory.

At the moment you will need to use two scripts or this workaround

```
import subprocess

print("Starting interactive visualization...")
p1 = subprocess.Popen(["python", "vis1.py")

p1.wait()
print("Starting rendering process...")
subprocess.run(["python", "vis2.py")
```

vis1.py should include:
```
keyframe_camera = vis.KeyframeCamera()
visualizer = vis.Visualizer(particles=particle_list,
                output_folder=out_folder,
                keyframe_camera=keyframe_camera)
visualizer.run_visualization()
```

vis2.py should include :
```
keyframe_camera = vis.KeyframeCamera(view_matrices_path='./output_folder/interpolated_view_matrices.npy')

visualizer2 = vis.Headless_Visualizer(particles=particle_list, 
                                    output_folder=out_folder,
                                    camera=keyframe_camera,)

visualizer2.render_visualization()
```

### Trajectory Camera

Allows defining a subclass ```trajectory``` object, which provides a trajectory the camera should follow through a scene.
Can be used as
```
trajectory_camera = vis.TrajectoryCamera(total_frames=number_of_frames,
						trajectory=trajectory_object)

```


#### Trajectories
Provides methods that return center, eye and up vector depending on the frame index depending on the total number of frames to create smooth trajectories the camera can follow.
At the moment the following trajectories are implemented
- Circular Trajectory
Creates a circular trajectory with ```radius```around the ```axis```-axis focusing on the ```center```. Optionally, the trajectory can be looped, ping-ponged or limited on a certain angle range. The ```frames_per_rotation``` variable defines the speed of the rotation.
An example would be:
```
circular_trajectory = vis.CircularTrajectory(total_frames=number_of_frames, 
							center=np.array([Lx/2, Ly/2, 0]), 
							radius=200,
							frames_per_rotation=20, 
							axis='y', 
							clockwise=True, 
							loop=False, 
							angle_range=(-np.pi/4, np.pi/4), 
							ping_pong=True)
```

- Zooming Trajectory
Creates a zooming in or zooming out trajectory from ```center``` and ```initial_eye``` based on the sign of ```zoom_distance```. The ```frames_per_zoom``` argument defines the ratio during which zooming happens.
An example would be:
```
zoom_trajectory = vis.ZoomingTrajectory(total_frames=number_of_frames,
				center=np.array([Lx/2, Ly/2, 0]),
				initial_eye=np.array([Lx/2, Ly/2, 200]), 
				zoom_distance = 150, 
				frames_per_zoom=1.0)
```

jdrotleff committed:
Fully renamed CircularCameraTrajectory to CircularTrajectory

jdrotleff committed:
Renamed BaseCameraTrajectory to BaseTrajectory

jpd-de commented:
I forgot to adapt the tests to some changes I made in the keyframe_camera.py... Oops.

jdrotleff committed:
Added missing return in keyframe_camera.py

SamTov commented on CI/test_files/keyframe_camera/interpolated_view_matrices.npy:
Do we need this?

jpd-de replied:
Yes, if we want to test the load_view_matrices method in the `keyframe camera`.

SamTov commented:
Can we clean up the UI. Why do we need two scripts?

SamTov commented on znvis/__init__.py:
```diff
@@ -24,6 +24,13 @@
 from znvis.bounding_objects.bounding_box import BoundingBox
+from znvis.cameras.camera import Camera
+from znvis.cameras.keyframe_camera import KeyframeCamera
+from znvis.cameras.particle_following_camera import ParticleFollowingCamera
```
I would prefer to have these wrapped under the camera object. So can you move these classes to that init?

jpd-de replied:
Yes

SamTov commented on znvis/cameras/__init__.py:
```diff
@@ -0,0 +1,34 @@
+from .trajectory_camera import TrajectoryCamera
+
+__all__ = [
+    "Camera",
```
These should not be strings but `cls.__name__`

SamTov commented on znvis/cameras/camera.py:
```diff
@@ -0,0 +1,264 @@
+        up : np.ndarray shape=(3,)
+                The up vector of the camera.
+        ↑ (up)
+        |
```
This is a very cool doc string

jpd-de replied:
Thanks!

SamTov commented on znvis/cameras/trajectories/__init__.py:
```diff
@@ -0,0 +1,32 @@
+from .zooming_trajectory import ZoomingTrajectory
+
+__all__ = [
+    "BaseTrajectory",
```
Same deal here, please remove the strings.

SamTov commented on znvis/cameras/__init__.py:
```diff
@@ -0,0 +1,34 @@
+Package for the ZnVis Cameras.
+"""
+
+from .camera import Camera
```
Please add the full import.

SamTov commented on znvis/cameras/keyframe_camera.py:
```diff
@@ -0,0 +1,277 @@
+        This function is called in the visualizer.py to initialize the
+        keyframe collection.
+        """
+        self.view_matrices_dictionary = {}
```
Why not just put it into the init?

SamTov commented on znvis/cameras/trajectories/__init__.py:
```diff
@@ -0,0 +1,32 @@
+Package for the ZnVis Cameras.
+"""
+
+from .base_trajectory import BaseTrajectory
```
Full import path

SamTov commented on znvis/cameras/trajectories/circular_trajectory.py:
```diff
@@ -0,0 +1,190 @@
+
+import numpy as np
+
+from .base_trajectory import BaseTrajectory
```
Full import pipelines.

SamTov commented on znvis/cameras/keyframe_camera.py:
```diff
@@ -0,0 +1,277 @@
+    """
+    A class to produce an smooth camera trajectory based on manually
+    picked frames and camera settings pairs.
+    """
```
Explain the interpolation shortly.

jpd-de replied:
Sure.
At the moment, only linear interpolation with a smoothing via SVD is currently implemented.
This can lead to unexpected results if the camera positions are chosen too far apart. Be aware of this and pick the camera positions accordingly. Especially take care if you want to zoom in or out. However, this way you have total control over the camera trajectory and can create very nice custom animations.

SamTov commented on znvis/cameras/particle_following_camera.py:
```diff
@@ -0,0 +1,84 @@
+                The positions of the particles in the simulation.
+        camera_particle_vector : np.ndarray shape=(3,)
+                The distance vector between the camera and the particle.
+        camera_up_vector : np.ndarray shape=(3,)
```
Make `(n_frames, 3)` or just `(3,)`

SamTov replied:
and make the comment that the dynamic version is how you turn on particle view.

jpd-de replied:
Done.
Three possible cases are considered:
    1. The camera follows the particle with a fixed distance vector.  Activated by passing a single vector of shape (3,) to the `camera_particle_vector` parameter.
    2. The camera follows the particle with a dynamic distance vector given by the user. Activated by passing an array of shape (n_frames, 3) to the `camera_particle_vector` parameter.
    3. The camera follows the particle while always looking at the same direction as the particle. Activated by passing the `particle_directions` parameter.

SamTov reviewed (changes requested):
Please take a look at the comments.

jdrotleff committed:
Adapted the PR to latest feedback

- Changed the init structure
- Changed strings in inits to cls.__names__
- Reorganized the BaseCamera class and the StaticCamera
- Moved `self.view_matrices_dict` initialization into the init of the keyframe camera
- Expanded the docstring about the interpolation method
- Clarified the `ParticleFollowingCamera` class

jdrotleff committed:
Fix the variable particle number features

Also added the related tests.

SamTov committed:
Merge branch 'main' into headless-renderer

jdrotleff committed:
Correcting tests

jdrotleff committed:
Implement coderabbitai feedback

Adapted tests to new code.

jdrotleff committed:
Adapted trajectory_camera

jdrotleff committed:
Reworked the headless renderer and keyframe camera

Seems like with an open3d version newer than 0.17.0 several instances of the headless renderer and the visualizer can be run from the same script.
Adapted the functionalities to that.
Fixed several issues highlighted by the coderabbitai
Reworked the interface of the keyframe camera to make usage more intuitive.

jdrotleff committed:
Handle empty frames to avoid IndexError

Also normalized director treatment when position is a list.

jdrotleff committed:
Adapted test_particle.py to new ValueError messages

jdrotleff committed:
Adapted to coderabbitai suggestions + new VideoManager class

This lets video creation be handled similar for the headless_renderer and the visualizer.

SamTov commented on CI/unit_tests/video/test_video_manager.py:
```diff
@@ -0,0 +1,128 @@
+        assert self.manager.validate_video_format("mp4") == "mp4"
+        assert self.manager.validate_video_format("xyz") == "mp4"  # fallback
+
+    def test_create_video_from_frames_mp4(self):
+        video_path = self.manager.create_video_from_frames(
+            frame_folder=self.frame_folder,
+            video_name="test_video",
+            video_format="mp4",
+            keep_frames=False,
+            frame_pattern="*.png",
+        )
```
Frame folder is probably just the path. Do the later tests need the actual content of the folder? If not, it can stay False

jdrotleff committed:
Added BaseVisualizer class

Also fixed several bugs and best-practice suggestions

jdrotleff committed:
Fixed bug in StaticCamera

Test was incompatible with _frame_index

jdrotleff committed:
Input validation for static_camera

jdrotleff committed:
Merge branch 'upstream/main' into headless-renderer

jdrotleff committed:
Added test for the headless_renderer.py

jdrotleff committed:
Added TearDownClass in headless renderer test

jdrotleff commented:
Note: There is room for optimizing memory handling for large trajectories. I observed memory overflow when rendering.

jdrotleff committed:
Enable lazy mesh generation for headless rendering to reduce memory usage

jdrotleff committed:
Add robust headless parallel rendering with spawn-safe workers and cleanup

SamTov commented:
Is this stuff all ready to go?

jdrotleff commented:
No, I still need to optimize the parallelization. It's currently all going to the GPU, not using CPU cores.

jdrotleff committed:
Refactor headless parallel rendering and simplify frame selection

jdrotleff committed:
Remove outdated test

jdrotleff committed:
expand docstrings

jdrotleff commented:
### Latest commits include

1. Headless rendering became lazy/memory-safe by adding lazy mesh construction. Previously, all meshes were created and stored in memory.
- Meshes for `Particle`/`VectorField` can now be generated per frame on demand.
- Avoids preloading all meshes into RAM.

2. New robust parallel headless renderer
- Added [`_parallel_render.py`](/tikhome/jdrotleff/Packages/ZnVis/znvis/visualizer/_parallel_render.py) and worker entrypoint [`_parallel_render_worker.py`](/tikhome/jdrotleff/Packages/ZnVis/znvis/visualizer/_parallel_render_worker.py).
- Uses isolated subprocesses + serialized shared state + JSON signaling.
- Per-worker GPU pinning via `CUDA_VISIBLE_DEVICES`.
- Failure paths fall back to serial rendering.

3. API cleanup/simplification
- `parallel_render_workers=None` means one worker per visible GPU. This is the default.
- `parallel_render_workers > n_visible_gpus` is supported but explicitly marked experimental (warning).

4. Frame selection and resumability in headless render
- `render_visualization(frame_indices=...)`
- `render_visualization(frame_range=(start, stop))`
- `skip_existing_frames=True` default for resumable rendering.
- Added `create_video_from_frames()` helper.

5. GPU detection hardening
- `_detect_available_gpu_devices()` uses `CUDA_VISIBLE_DEVICES` + Dr.Jit detection.

6. Tests/docs updated
- Added/expanded parallel render tests and headless selection/resume tests.
- Expanded docstrings and getting-started docs around headless/parallel behavior.
- Added lint/format config updates (`.isort.cfg`, pre-commit tweak).

### What's not yet included but should be added:
Lazy mesh-construction in for interactive visualizations.

jdrotleff committed:
Remove unnecessary variable assignment

jdrotleff committed:
Hint to use the HeadlessRenderer when running a rendering from visualizer

jdrotleff committed:
fix: avoid pickling TestCase method

jdrotleff committed:
feature: Add lazy mesh loading with rolling window cache

jdrotleff committed:
Remove debugging methods

jdrotleff commented:
### Feature: Implement lazy mesh loading
**Summary:**
The last commit introduces a background caching architecture to lazily load and evict meshes during visualization.
This replaces the default eager-loading approach that could lead to OOM errors for larger trajectories.

The system is divided in three components:
1. **MeshCachePolicy**: Evaluates the current playback state (position, direction, speed) and returns a TargetWindow—an explicit set of frames to retain and a prioritized list of frames to build.
2. **MeshFrameCache**: A byte-aware LRU cache that stores generated Open3D meshes and evicts the least recently used entries when max_bytes is exceeded.
3. **MeshCacheManager (Background Worker)**: Runs a single thread that strictly executes the TargetWindow instructions, staying always ahead of the UI thread.

More details:
* Completeness Map: Trajectory readiness checks avoid O(N) scanning across particles. A `complete_frames` set, guarded by a threading lock, acts as the single source of truth, allowing both the UI and background threads to check available frames.
* Bounded Chunk Execution: The background worker strictly processes missing frames in fixed batches (default `chunk_size = 10`). By yielding execution after each chunk, the thread continuously checks for UI updates, allowing it to instantly pivot if the user reverses direction or scrubs the timeline without locking up the CPU.
* Playback-Speed compatible: The policy explicitly factors in playback speed (frame_step). During high-speed playback (e.g., 8x), the cache targets [0, 8, 16, 24...], explicitly skipping frames to prevent inefficient prefetching.
* Safety bounds for increasing number of particles: The background prefetch horizon is dynamically coupled to physical RAM limits. The retention window organically shrinks to match reality, if heavy meshes are loaded.

### How to use
```
visualizer = Visualizer(
    ....
    lazy_mesh_loading=True,
    mesh_cache_max_gb=2.5,  # Cap background mesh RAM at 2.5 GB
    future_fraction=0.5    # Dedicate 50% of the cache to prefetching forward, 50% for keeping frames if you want to switch play direction often.
)
```

#### Minor changes
* Addressed a bug where backward playing was not correctly updating frames.
* Refactored parallel and cache into subpackages in visualizer. Could also live elsewhere

jdrotleff committed:
Isolate ZnVis parallel wprker json stream from stdout and stderr

jdrotleff committed:
Linting

SamTov reviewed (changes requested):
It's really close. I don't like the second layer of directories for `cache`, `parallel`, and `trajectory`. Can you bring those up a layer? They could all be re-purposed for other things potentially, they don't need to be buried. If you could also give them slightly clearer names, like `camera_trajectory`, `mesh cache`, and `parallel_render`, that would be great. Otherwise, LGTM.

jdrotleff committed:
Renaming and refactoring of cache, parallel and trajectory

cache -> mesh_cache
parallel -> parallel_render
trajectory -> camera_trajectory

jdrotleff commented:
Fair point, I renamed and restructured `cache` -> `mesh_cache`, `parallel` -> `render_parallel`, and `trajectory` -> `camera_trajectory`

SamTov reviewed (approved):
LGTM thanks for the updates.
